pyro/ops/jit.py

```diff
@@ -0,0 +1,79 @@
+import weakref
+import torch
+
+import pyro
+import pyro.poutine as poutine
+
+
+def compile(**jit_options):
+    """
+    Drop-in replacement for :func:`torch.jit.compile` that works with
+    Pyro functions that call :func:`pyro.param`.
+
+    The actual compilation artifact is stored in ``self.compiled``.
+    Call diagnostic methods on this attribute.
+
+    Example::
+
+        def model(x):
+            scale = pyro.param("scale", torch.tensor(0.5), constraint=constraints.positive)
+            return pyro.sample("y", dist.Normal(x, scale))
+
+        @pyro.ops.jit.compile(nderivs=1)
+        def model_log_prob_fn(x, y):
+            cond_model = pyro.condition(model, data={"y": y})
+            tr = pyro.poutine.trace(cond_model).get_trace(x)
+            return tr.log_prob_sum()
+    """
+    class wrapper(object):
```

Comment:
I would recommend defining this big class outside of `compile`:
```py
class CompiledFunction(object):  # formerly class wrapper
    def __init__(self, fn, jit_options):
        ...
```
and then adding decorator magic in the `compile` helper, something like
```
def compile(fn=None, **jit_options):
    if fn is None:
        return lambda fn: compile(fn, **jit_options)
    return CompiledFunction(fn, jit_options) 
```